csharp/src/Google.Protobuf/UnknownFieldSet.cs

```diff
@@ -0,0 +1,340 @@
+#region Copyright notice and license
+// Protocol Buffers - Google's data interchange format
+// Copyright 2015 Google Inc.  All rights reserved.
+// https://developers.google.com/protocol-buffers/
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+//     * Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//     * Redistributions in binary form must reproduce the above
+// copyright notice, this list of conditions and the following disclaimer
+// in the documentation and/or other materials provided with the
+// distribution.
+//     * Neither the name of Google Inc. nor the names of its
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using Google.Protobuf.Reflection;
+
+namespace Google.Protobuf
+{
+	/// <summary>
+	/// Used to keep track of fields which were seen when parsing a protocol message
+	/// but whose field numbers or types are unrecognized. This most frequently
+	/// occurs when new fields are added to a message type and then messages containing
+	/// those fields are read by old software that was built before the new types were
+	/// added.
+	///
+	/// Most users will never need to use this class directly.
+	/// </summary>
+	public sealed partial class UnknownFieldSet : IMessage
+	{
+		private static readonly UnknownFieldSet defaultInstance =
+				new UnknownFieldSet();
+		private readonly IDictionary<int, UnknownField> fields;
+
+		/// <summary>
+		/// Creates a new UnknownFieldSet.
+		/// </summary>
+		public UnknownFieldSet()
+		{
+			this.fields = new Dictionary<int, UnknownField>();
+		}
+
+		/// <summary>
+		/// Gets the default UnknownFieldSet instance.
```

Comment:
We can't sensibly do this while the type is mutable. Rather than talk about this in individual comments, I'll write an overall comment later.

Reply:
Have remove it and use null instead